src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -34,24 +34,44 @@ public UdfComplexTypesTests(SparkFixture fixture)
         [Fact]
         public void TestUdfWithArrayType()
         {
-            // UDF with ArrayType throws a following exception:
-            // [] [] [Error] [TaskRunner] [0] ProcessStream() failed with exception: System.InvalidCastException: Unable to cast object of type 'System.Collections.ArrayList' to type 'System.Int32[]'.
-            //  at Microsoft.Spark.Sql.PicklingUdfWrapper`2.Execute(Int32 splitIndex, Object[] input, Int32[] argOffsets) in Microsoft.Spark\Sql\PicklingUdfWrapper.cs:line 44
-            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
-            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
-            Func<Column, Column> udf = Udf<int[], string>(array => string.Join(',', array));
-            Assert.Throws<Exception>(() => _df.Select(udf(_df["ids"])).Show());
-
-            // Currently, there is a workaround to support ArrayType using ArrayList.
-            Func<Column, Column> workingUdf = Udf<ArrayList, string>(
+            // Int Array.
+            {
+                Func<Column, Column> udf = Udf<int[], string>(
                 array => string.Join(',', array.ToArray()));
 
-            Row[] rows = _df.Select(workingUdf(_df["ids"])).Collect().ToArray();
-            Assert.Equal(3, rows.Length);
+                Row[] rows = _df.Select(udf(_df["ids"])).Collect().ToArray();
+                Assert.Equal(3, rows.Length);
+
+                var expected = new string[] { "1", "3,5", "2,4" };
+                string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();
+                Assert.Equal(expected, rowsToArray);
+            }
+
+            // Double Array Array.
+            {
+                Func<Column, Column> udf = Udf<double[][], double>(
+                doubleArrArr => doubleArrArr[0][0] + 100);
 
-            var expected = new string[] { "1", "3,5", "2,4" };
-            string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();
-            Assert.Equal(expected, rowsToArray);
+                Row[] rows = _df.Select(udf(_df["scores"])).Collect().ToArray();
+                Assert.Equal(3, rows.Length);
+
+                var expected = new string[] { "101", "101.1", "101.2" };
+                string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();
+                Assert.Equal(expected, rowsToArray);
+            }
+
+            // Array of Rows.
+            {
+                Func<Column, Column> udf = Udf<Row[], string>(
+                rowArr => rowArr[0].GetAs<string>("city"));

```

Comment:
Can you also test adding in an array as a field to the row ?

Reply:
Sure, will add the test tonight.

Reply:
Added this test.